### PR TITLE
Add support for Python 3.4/3.5 and PyPy.

### PR DESCRIPTION
**Pull request by jamadden**
This was almost entirely test changes.

There was one syntax change, one import change, and one
generator-to-list change. Plus some whitespace cleanup.

There is one test in utils.txt that was broken on both Python 2 and
Python 3; I kluged up a "fix" but the functionality seems broken in all
versions currently.

### CODE REVIEW

File: src/z3c/autoinclude/dependency.txt [hunk 1]
```diff
@@ -86,8 +86,8 @@ that it depends on that have ``configure.zcml`` or ``meta.zcml``.
 Note that the individual lists within ``includableInfo`` preserve the
 package order defined in ``setup.py``::
 
-    >>> a_include_finder.includableInfo(['configure.zcml', 'meta.zcml'])
-    {'configure.zcml': ['b.c'], 'meta.zcml': ['z3c.autoinclude', 'testdirective']}
+    >>> list(sorted(list(a_include_finder.includableInfo(['configure.zcml', 'meta.zcml']).items())))
```
Comment (mgedmin) [inline src/z3c/autoinclude/dependency.txt]: You can simplify that to `list(sorted(a_include_finder....items()))` -- sorted() accepts any iterable.

Perhaps a slightly more readable alternative would be to `pprint.pprint(...)` the dict.

File: src/z3c/autoinclude/utils.py [hunk 2]
```diff
@@ -89,9 +90,9 @@ def distributionForDottedName(package_dottedname):
 
     if len(valid_dists_for_package) == 0:
         raise LookupError("No distributions found for package `%s`; are you sure it is importable?" % package_dottedname)
-    
+
     if len(valid_dists_for_package) > 1:
-        non_namespaced_dists = filter(lambda x: len(x[1]) is 0, valid_dists_for_package)
+        non_namespaced_dists = list(filter(lambda x: len(x[1]) is 0, valid_dists_for_package))
```
Comment (mgedmin) [inline src/z3c/autoinclude/utils.py]: Ugh.  I suppose it works, but rewriting as a comprehension would be better.  If you can figure out what is going on here ;)

File: src/z3c/autoinclude/utils.txt [hunk 3]
```diff
@@ -6,13 +6,13 @@ distributionForPackage is a function that takes a module object
 and returns the setuptools distribution object that contains
 the module.
 
-It should find the correct distribution for a package whose namespace
-is extended by other packages in the environment::
+It *should* find the correct distribution for a package whose namespace
+is extended by other packages in the environment. However it does *not*::
```
Comment (mgedmin) [inline src/z3c/autoinclude/utils.txt]: I don't know how I feel about encoding incorrect behavior in a test.

File: src/z3c/autoinclude/dependency.txt (hunk 1 above)
Comment (jamadden) [inline src/z3c/autoinclude/dependency.txt]: I ran into problems with `pprint` being switched to `compact=False` in Python 3.4 in another place and worked around it using a `width` parameter, but that's pretty fragile, it seemed to me. It also had issues with the unicode-prefix normalizer. So I wasn't feeling the pprint love :) Plus, I'm not sure...is pprint guaranteed to sort a dictionary?

File: src/z3c/autoinclude/utils.txt (hunk 3 above)
Comment (jamadden) [inline src/z3c/autoinclude/utils.txt]: I know 😦 This test used to pass, apparently, but no longer does. This suggests a change in external dependencies of some sort. But the code being tested is pretty opaque and I didn't want to mess with it.

I can/will try to debug what's going wrong, but that seemed like too much of a change for a simple port PR.

File: src/z3c/autoinclude/utils.py (hunk 2 above)
Comment (jamadden) [inline src/z3c/autoinclude/utils.py]: Good idea. Done.

File: src/z3c/autoinclude/dependency.txt (hunk 1 above)
Comment (jamadden) [inline src/z3c/autoinclude/dependency.txt]: Well, regardless of guarantees, it seems to work :)

File: src/z3c/autoinclude/utils.txt (hunk 3 above)
Comment (jamadden) [inline src/z3c/autoinclude/utils.txt]: #4

**Comment by jamadden** [comment]
Whoops, hold on, the cleanup introduced a bug....which isn't caught by the test cases.

**Comment by jamadden** [comment]
Ok, I debugged through the test failure. We're just making an arbitrary choice, and the comments claim that an arbitrary choice is perfectly acceptable because the elements are equivalent. So it comes down to the order in which `find_distributions` returns things, and the order in which things are on sys.path. 

So for the sake of test reproducibility, I added a small preference to try to return distributions that match the package name in simple cases (I think #2 winds up doing something similar).

File: src/z3c/autoinclude/dependency.txt (hunk 1 above)
Comment (mgedmin) [inline src/z3c/autoinclude/dependency.txt]: pprint() sorts the dictionary, unless you're on python 2.4 and len(repr(the_dictionary)) < 70 or something like that.  zope.testing used to have a fork of pprint.pprint to fix that one little buglet.

The pprint changes in Python 3.4 are new to me and sound like a valid reason not to use pprint (or doctests). ;)

File: src/z3c/autoinclude/api.py [hunk 4]
```diff
@@ -4,15 +4,15 @@
...
 def enable_dependencies():
     del os.environ[DEP_KEY]
 
 def plugins_disabled():
-    return os.environ.has_key(PLUGIN_KEY)
+    return PLUGIN_KEY in os.environ
 def disable_plugins():
     os.environ[PLUGIN_KEY] = 'True'
-def enable_dependencies():
+def enable_plugins():
```
Comment (mgedmin) [inline src/z3c/autoinclude/api.py]: Nice catch!

File: src/z3c/autoinclude/tests/tests.py [hunk 5]
```diff
@@ -68,12 +71,19 @@ def testTearDown(test):
 
 IGNORECASE = doctest.register_optionflag('IGNORECASE')
 
-class IgnoreCaseChecker(doctest.OutputChecker):
+class IgnoreCaseChecker(renormalizing.RENormalizing, object):
```
Comment (mgedmin) [inline src/z3c/autoinclude/tests/tests.py]: I'm afraid to ask.  ... Is ... is RENormalizing an old-style class?  O.o

File: src/z3c/autoinclude/utils.py [hunk 6]
```diff
@@ -89,12 +92,20 @@ def distributionForDottedName(package_dottedname):
 
     if len(valid_dists_for_package) == 0:
         raise LookupError("No distributions found for package `%s`; are you sure it is importable?" % package_dottedname)
-    
+
     if len(valid_dists_for_package) > 1:
-        non_namespaced_dists = filter(lambda x: len(x[1]) is 0, valid_dists_for_package)
+        non_namespaced_dists = [x for x in valid_dists_for_package if len(x[1]) is 0]
```
Comment (mgedmin) [inline src/z3c/autoinclude/utils.py]: The horridness of that filter/lambda monstrocity originally obscured the `is 0` comparison from my sight, which can we *please* replace with `== 0`?  Or with `if not x[1]`?

Ideally I'd like to see `[(x, y) for x, y in valid_dists_for_package if not y]`, but with `x` and `y` replaced with meaningful names.

File: src/z3c/autoinclude/tests/tests.py (hunk 5 above)
Comment (jamadden) [inline src/z3c/autoinclude/tests/tests.py]: Yes 😦

File: src/z3c/autoinclude/utils.txt [hunk 7]
```diff
@@ -27,7 +27,7 @@ by having been extended by nested packages) it should find a package::
 
     >>> import enolp
     >>> distributionForPackage(enolp)
-    enolp.ppa.foo 0.1 (...enolp.ppa.foo-0.1...egg)
+    enolp.ppa.bar 0.1 (...enolp.ppa.bar-0.1...egg)
```
Comment (mgedmin) [inline src/z3c/autoinclude/utils.txt]: Is this deterministic now?  If not maybe we can pick the 1st one alphabetically?

File: src/z3c/autoinclude/utils.py (hunk 6 above)
Comment (jamadden) [inline src/z3c/autoinclude/utils.py]: Yikes, I missed the equality/is thing too.

File: src/z3c/autoinclude/utils.txt (hunk 7 above)
Comment (jamadden) [inline src/z3c/autoinclude/utils.txt]: Good point. It seems like ignoring the order that entries are on sys.path could have consequences, but I'll trust the comment that says they're equivalent.

Done.

**Comment by mgedmin** [comment]
LGTM!

Awesome fixes BTW!

File: src/z3c/autoinclude/api.py (hunk 4 above)
Comment (jamadden) [inline src/z3c/autoinclude/api.py]: pylint FTW!

**Comment by jamadden** [comment]
Perhaps this warrants a PyPI release? (I don't need it now so there's no rush on my account, but eventually my projects will need it.)

**Comment by mgedmin** [comment]
PyPI says:

> Package Index Owner: faassen, ejucovy, janjaapdriessen, rmarianski, hannosch, davisagli, esteele, timo, plone, thefunny42, jw

so, uh, ping @faassen @ejucovy @janjaapdriessen @hannosch @davisagli @esteele @thefunny42?  Can we please have a release of z3c.autoinclude (and/or `mgedmin` in the PyPI owners list)?

**Comment by davisagli** [comment]
`mgedmin` and `jamadden` are now owners on PyPI

**Comment by jamadden** [comment]
That was unexpected, thanks! It looks like we're zest.releaser compatible, so I'll try my hand at pushing a release.

**Pull request merged**